pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/SegmentPreprocessThrottler.java

```diff
@@ -0,0 +1,139 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.utils;
+
+import com.google.common.annotations.VisibleForTesting;
+import java.util.Map;
+import org.apache.pinot.common.concurrency.AdjustableSemaphore;
+import org.apache.pinot.spi.config.provider.PinotClusterConfigChangeListener;
+import org.apache.pinot.spi.utils.CommonConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Used to throttle the total concurrent index rebuilds that can happen on a given Pinot server.
+ * Code paths that do no need to rebuild the index or which don't happen on the server need not utilize this throttler.
+ */
+public class SegmentPreprocessThrottler implements PinotClusterConfigChangeListener {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentPreprocessThrottler.class);
+
+  /**
+   * If _maxPreprocessConcurrency is <= 0, it means that the cluster is not configured to limit the number of
+   * index rebuilds that can be executed concurrently. This class sets the semaphore permit to be a large value
+   * in this scenario.
+   */
+  private int _maxPreprocessConcurrency;
+  private boolean _relaxThrottling;
+  private volatile AdjustableSemaphore _semaphore;
+
+  public SegmentPreprocessThrottler(int maxPreprocessConcurrency, boolean relaxThrottling) {
+    LOGGER.info("Initializing SegmentPreprocessThrottler, maxPreprocessConcurrency: {}, relaxThrottling: {}",
+        maxPreprocessConcurrency, relaxThrottling);
+    if (maxPreprocessConcurrency <= 0) {
+      int defaultSegmentPreprocessParallelism =
```

Comment:
Let's directly throw exception here to not let server starts. Using default for invalid config could cause very confusing behavior

Reply:
sure, i've made these changes. for handling ZK cluster config changes to make this -ve or 0, I've just logged a warning and don't make the change rather than throwing an exception from the handler. Let me know if you think this should be handled differently for that scenario.

Reply:
Sounds good. We don't want to throw exception from the handler because it might cause other handler from running